src/Microsoft.OData.Core/JsonLight/ODataJsonLightServiceDocumentDeserializer.cs

```diff
@@ -312,48 +304,333 @@ private ODataServiceDocumentElement ReadServiceDocumentElement(PropertyAndAnnota
 
-            this.JsonReader.ReadEndObject();
+            await this.JsonReader.ReadEndObjectAsync()
+                .ConfigureAwait(false);
+
+            return serviceDocumentElement;
+        }
+
+        /// <summary>
+        /// Creates a <see cref="ODataServiceDocumentElement"/>.
+        /// </summary>
+        /// <param name="kind">Property values for the "kind" property.</param>
+        /// <returns>A <see cref="ODataServiceDocumentElement"/> instance.</returns>
+        private static ODataServiceDocumentElement CreateServiceDocumentElement(string[] kind)
+        {
+            // If not specified its an entity set.
+            if (kind[0] == null)
+            {
+                return new ODataEntitySetInfo();
+            }
+
+            ODataServiceDocumentElement serviceDocumentElement = null;
+
+            if (kind[0].Equals(JsonLightConstants.ServiceDocumentEntitySetKindName, StringComparison.Ordinal))
+            {
+                serviceDocumentElement = new ODataEntitySetInfo();
+            }
+            else if (kind[0].Equals(JsonLightConstants.ServiceDocumentFunctionImportKindName, StringComparison.Ordinal))
+            {
+                serviceDocumentElement = new ODataFunctionImportInfo();
+            }
+            else if (kind[0].Equals(JsonLightConstants.ServiceDocumentSingletonKindName, StringComparison.Ordinal))
+            {
+                serviceDocumentElement = new ODataSingletonInfo();
+            }
 
             return serviceDocumentElement;
         }
+
+        /// <summary>
+        /// Validates that the specified property in the service document element is not repeated.
+        /// </summary>
+        /// <param name="property">The property.</param>
+        /// <param name="propertyValues">Property values for the specified property.</param>
+        private static void ValidateServiceDocumentElementHasNoRepeatedProperty(string property, string[] propertyValues)
+        {
+            Debug.Assert(propertyValues != null && propertyValues.Length > 0, $"{nameof(propertyValues)} != null && {nameof(propertyValues)}.Length > 0");
+
+            if (propertyValues[0] != null)
+            {
+                throw new ODataException(Strings.ODataJsonLightServiceDocumentDeserializer_DuplicatePropertiesInServiceDocumentElement(property));
+            }
+        }
+
+        /// <summary>
+        /// Validates that the specified required property in the service document element has been read and value is not null.
+        /// </summary>
+        /// <param name="property">The property.</param>
+        /// <param name="propertyValues">Property values for the specified property.</param>
+        private static void ValidateServiceDocumentElementHasRequiredProperty(string property, string[] propertyValues)
+        {
+            Debug.Assert(propertyValues != null && propertyValues.Length > 0, $"{nameof(propertyValues)} != null && {nameof(propertyValues)}.Length > 0");
+

```

Comment:
Suggestion: 
``` 
            Debug.Assert(propertyValues != null && propertyValues.Length > 0, $"{property} != null && {property}.Length > 0");
````

Reply:
I have a string array `string[] propertyValues`
This:
```csharp
$"{propertyValues} != null && {propertyValues}.Length > 0"
```
gives me `" != null && .Length > 0"`
Which is not what I'm looking for.

On the other hand,
```csharp
$"{nameof(propertyValues)} != null && {nameof(propertyValues)}.Length > 0"
```
gives me `"propertyValues != null && propertyValues.Length > 0"`